### PR TITLE
GraphQL schema-first: handle persistence backends that default to SAI indexes

### PR DESCRIPTION
**Pull request by olim7t**

### CODE REVIEW

File: graphqlapi/src/main/java/io/stargate/graphql/web/HttpAwareContext.java [hunk 1]
```diff
@@ -41,9 +43,10 @@
   // For more information.
   private final BatchContext batchContext = new BatchContext();
 
-  public HttpAwareContext(HttpServletRequest request) {
+  public HttpAwareContext(HttpServletRequest request, Persistence persistence) {
```
Comment (olim7t) [inline graphqlapi/src/main/java/io/stargate/graphql/web/HttpAwareContext.java:46]: This is more convenient than modifying all the fetcher constructors.
I think we should do the same for existing constructor arguments, see #945.

File: graphqlapi/src/main/java/io/stargate/graphql/schema/graphqlfirst/processor/IndexModelBuilder.java [hunk 2]
```diff
@@ -59,6 +59,10 @@
   IndexModel build() {
     Optional<String> indexClass =
         DirectiveHelper.getStringArgument(cqlIndexDirective, "class", context);
+    // Some persistence backends default to SAI when no class is specified:
+    if (!indexClass.isPresent() && !context.defaultsToRegularIndexes()) {
+      indexClass = Optional.of("org.apache.cassandra.index.sai.StorageAttachedIndex");
```
Comment (dimas-b) [inline graphqlapi/src/main/java/io/stargate/graphql/schema/graphqlfirst/processor/IndexModelBuilder.java:64]: Should we also check `Persistence.supportsSAI)` in this case?

Although, in practice we do not have `Persistence` implementations that support neither 2i nor SAI, we could probably produce a nicer error message if it ever happens in the future. WDYT?

Comment (olim7t) [reply graphqlapi/src/main/java/io/stargate/graphql/schema/graphqlfirst/processor/IndexModelBuilder.java:64]: 👍 you're right, I'll make the change.

Comment (tomekl007) [reply graphqlapi/src/main/java/io/stargate/graphql/schema/graphqlfirst/processor/IndexModelBuilder.java:64]: Also, maybe log the info, i.e:
`log.info("The index is not set, defaulting to sai")`
?

Comment (olim7t) [reply graphqlapi/src/main/java/io/stargate/graphql/schema/graphqlfirst/processor/IndexModelBuilder.java:64]: 👍 I added a processing message

**Review by dimas-b — APPROVED** [review]

File: graphqlapi/src/main/java/io/stargate/graphql/schema/graphqlfirst/processor/IndexModelBuilder.java (hunk 2 above)
Comment (dimas-b) [reply graphqlapi/src/main/java/io/stargate/graphql/schema/graphqlfirst/processor/IndexModelBuilder.java:82]: Maybe add a warning to the context, similar to `invalidMapping(...)` but a warning if we default to SAI? Would it be helpful to the end user?

**Review by dougwettlaufer — APPROVED** [review]

File: graphqlapi/src/main/java/io/stargate/graphql/schema/graphqlfirst/processor/IndexModelBuilder.java (hunk 2 above)
Comment (olim7t) [reply graphqlapi/src/main/java/io/stargate/graphql/schema/graphqlfirst/processor/IndexModelBuilder.java:82]: Yeah that's pretty much what `info()` does. The message will be visible to the user if they select `logs` in the `deploySchema` response.

**Review by mpenick — APPROVED** [review]

File: graphqlapi/src/main/java/io/stargate/graphql/schema/graphqlfirst/processor/FieldModelBuilder.java [hunk 3]
```diff
@@ -319,9 +319,15 @@ private static boolean isCompatible(Column.ColumnType type1, Column.ColumnType t
         throw SkipException.INSTANCE;
       }
     }
-    return cqlIndexDirective.map(
-        d ->
-            new IndexModelBuilder(d, parentCqlName, cqlName, cqlType, messagePrefix, context)
-                .build());
+    return cqlIndexDirective.flatMap(
```
Comment (mpenick) [inline graphqlapi/src/main/java/io/stargate/graphql/schema/graphqlfirst/processor/FieldModelBuilder.java:322]: Why not move this up into the if statement above?

```java
if (cqlIndexDirective.isPresent()) {
   // ...
   return Optional.of(
     new IndexModelBuilder(d, parentCqlName, cqlName, cqlType, messagePrefix, context)
        .build());
}
```

Comment (mpenick) [reply graphqlapi/src/main/java/io/stargate/graphql/schema/graphqlfirst/processor/FieldModelBuilder.java:322]: Or move the other checks into the `flatMap()`? 

```java
return cqlIndexDirective.flatMap(
        d -> {
          try {

            if (isPk) {
              invalidMapping("%s: partition or clustering columns can't have an index", messagePrefix);
              throw SkipException.INSTANCE;
            }
            if (isUdtField) {
              invalidMapping("%s: UDT fields can't have an index", messagePrefix);
              throw SkipException.INSTANCE;
            }
            return Optional.of(
                new IndexModelBuilder(d, parentCqlName, cqlName, cqlType, messagePrefix, context)
                    .build());
          } catch (SkipException e) {
            return Optional.empty();
          }
        });
```

Comment (olim7t) [reply graphqlapi/src/main/java/io/stargate/graphql/schema/graphqlfirst/processor/FieldModelBuilder.java]: 👍 I applied your first suggestion.

**Review by mpenick — APPROVED** [review]

**Pull request merged**